### PR TITLE
Add sample_backward_noise_terms_with_state

### LINKED ISSUE #735: Error when using nonlinear function of state.in
I am trying to solve a problem that is dependent on a nonlinear function of both the incoming state (that sould be a parameter) and the uncertainty.

However, since both the uncertainty and the incoming state are treated as variables in JuMP I am getting the following error:

```
ERROR: MathOptInterface.UnsupportedConstraint{MathOptInterface.ScalarNonlinearFunction, MathOptInterface.GreaterThan{Float64}}: `MathOptInterface.ScalarNonlinearFunction`-in-`MathOptInterface.GreaterThan{Float64}` constraint is not supported by the model.
```
Here is a minimum working case based on one of SDDP.jl documentation examples, where we demonstrate the error on a trivial constraint. Is there a way around? 

```
using SDDP, Gurobi, Test

function fast_hydro_thermal()
    model = SDDP.LinearPolicyGraph(;
        stages = 2,
        upper_bound = 0.0,
        sense = :Max,
        optimizer = Gurobi.Optimizer,
    ) do sp, t
        @variable(sp, 0 <= x <= 8, SDDP.State, initial_value = 0.0)
        @variables(sp, begin
            y >= 0
            p >= 0
            ξ
        end)
        @constraints(sp, begin
            p + y >= 6
            x.out <= x.in - y + ξ
        end)
        RAINFALL = (t == 1 ? [6] : [2, 10])
        SDDP.parameterize(sp, RAINFALL) do ω
            return JuMP.fix(ξ, ω)
        end
        @stageobjective(sp, -5 * p)

        @constraint(sp, 1/x.in*exp(ξ) >= 0)

    end

   SDDP.train(model)
    return
end

fast_hydro_thermal()
```
For us this problem appears when we are implementing a specific stoachstic process for modeling the inflow inside the SDDP. Here are the equations for our case. 

```
    lower_bound = - (μ_stage) /  σ_stage - sum(ϕ[j]*inflow[i, end - j + 1] for j in 1:p)
    λ = (var_resid/lower_bound^2) + 1
    μ = 0.5 * log(var_resid/ (λ * (λ - 1)))
    σ = sqrt(log(λ))
    ϵ[i] = exp(ξ*σ+ μ) + lower_bound

    @constraint(subproblem, inflow_modeling[i = 1:n_hydro], inflow[i, 6].out == sum(ϕ[i, t][j]*inflow[i, end - j + 1].in for j in 1:p[i, t]) + ϵ[i])
```

Another way around would be to treat ϵ[i] as a parameter that is evaluated before solving the subproblem.

### PR DESCRIPTION
**Pull request by arthur-brigatto**
Closes #735.

### CODE REVIEW

**Comment by andrewrosemberg** [comment]
A simple test would be great to see if it works!

File: src/algorithm.jl [hunk 1]
```diff
@@ -660,7 +660,7 @@ function solve_all_children(
         end
         child_node = model[child.term]
         for noise in
-            sample_backward_noise_terms(backward_sampling_scheme, child_node)
+            sample_backward_noise_terms(backward_sampling_scheme, child_node, outgoing_state)
```
Comment (odow) [inline src/algorithm.jl:663]:
```suggestion
            sample_backward_noise_terms_with_state(
                backward_sampling_scheme,
                child_node,
                outgoing_state,
            )
```

File: src/plugins/backward_sampling_schemes.jl [hunk 2]
```diff
@@ -10,7 +10,7 @@ Backward sampler that returns all noises of the corresponding node.
 """
 struct CompleteSampler <: AbstractBackwardSamplingScheme end
 
-sample_backward_noise_terms(::CompleteSampler, node) = node.noise_terms
+sample_backward_noise_terms_with_state(::CompleteSampler, node::Node, state::Dict{Symbol,Float64}) = node.noise_terms
```
Comment (odow) [inline src/plugins/backward_sampling_schemes.jl:13]:
```suggestion
sample_backward_noise_terms(::CompleteSampler, node::Node) = node.noise_terms
```

File: src/plugins/backward_sampling_schemes.jl [hunk 3]
```diff
@@ -22,7 +22,7 @@ struct MonteCarloSampler <: AbstractBackwardSamplingScheme
     number_of_samples::Int
 end
 
-function sample_backward_noise_terms(sampler::MonteCarloSampler, node::Node)
+function sample_backward_noise_terms(sampler::MonteCarloSampler, node::Node, state::Dict{Symbol,Float64})
```
Comment (odow) [inline src/plugins/backward_sampling_schemes.jl:25]:
```suggestion
function sample_backward_noise_terms(sampler::MonteCarloSampler, node::Node)
```

File: src/plugins/backward_sampling_schemes.jl [hunk 4]
```diff
@@ -10,7 +10,7 @@ Backward sampler that returns all noises of the corresponding node.
 """
```
Comment (odow) [inline src/plugins/backward_sampling_schemes.jl:10]: I can't add a suggestion to the top of the file, but add the default implementation:
```julia
function sample_backward_noise_terms_with_state(
    sampler::AbstractBackwardSamplingScheme,
    node::Node,
    state::Dict{Symbol,Float64},
)
    return sample_backward_noise_terms(sampler, node)
end
```

Comment (odow) [reply src/plugins/backward_sampling_schemes.jl:10]: Actually, it should go here: https://github.com/odow/SDDP.jl/blob/8ed2e302c7212388ba3c4ec67c4737eaa44aeed5/src/plugins/headers.jl#L101-L120

**Comment by odow** [comment]
You can add a new test as a function just after this one:
https://github.com/odow/SDDP.jl/blob/8ed2e302c7212388ba3c4ec67c4737eaa44aeed5/test/plugins/backward_sampling_schemes.jl#L66-L90

File: src/plugins/headers.jl [hunk 5]
```diff
@@ -112,6 +112,7 @@ abstract type AbstractBackwardSamplingScheme end
     sample_backward_noise_terms(
         backward_sampling_scheme::AbstractBackwardSamplingScheme,
         node::Node{T},
+        state::Dict{Symbol,Float64},
```
Comment (odow) [inline src/plugins/headers.jl:115]: Okay, so:

I want two methods:

1. the existing `sample_backward_noise_terms`
2. a new `sample_backward_noise_terms_with_state`

as well as a fallback

```Julia
function sample_backward_noise_terms_with_state(
    sampler::AbstractBackwardSamplingScheme,
    node::Node,
    state::Dict{Symbol,Float64},
)
    return sample_backward_noise_terms(sampler, node)
end
```

The reason for this is to maintain backwards compatibility: we want all existing code to continue to work, but we want the ability for new code to use `sample_backward_noise_terms_with_state` instead.

Comment (odow) [inline src/plugins/headers.jl:115]:
```suggestion
```

Comment (arthur-brigatto) [reply src/plugins/headers.jl:115]: Alright, I had the wrong idea. I will correct it.

**Comment by arthur-brigatto** [comment]
> You can add a new test as a function just after this one:
> 
> https://github.com/odow/SDDP.jl/blob/8ed2e302c7212388ba3c4ec67c4737eaa44aeed5/test/plugins/backward_sampling_schemes.jl#L66-L90

Working on it!

File: src/plugins/headers.jl [hunk 6]
```diff
@@ -115,10 +115,28 @@ abstract type AbstractBackwardSamplingScheme end
...
 function sample_backward_noise_terms end
 
+"""
+    sample_backward_noise_terms_with_state(
+        sampler::AbstractBackwardSamplingScheme,
+        node::Node,
+        state::Dict{Symbol,Float64},
+    )::Vector{Noise}
+
+Returns a `Vector{Noise}` of noises sampled conditionally on the sate `state` using
```
Comment (odow) [inline src/plugins/headers.jl:129]:
```suggestion
Returns a `Vector{Noise}` of noises sampled conditionally on the `state` using
```

File: src/plugins/headers.jl [hunk 7]
```diff
@@ -115,10 +115,28 @@ abstract type AbstractBackwardSamplingScheme end
...
+`backward_sampling_scheme`.
```
Comment (odow) [inline src/plugins/headers.jl:130]:
```suggestion
`sampler`.
```

**Comment by arthur-brigatto** [comment]
@odow I have added a test function. The idea is: Firstly a sequence of 0.1 or 0.9 values for $\epsilon$ is sampled in the forward step. Then, the backward sampler sets $\epsilon = 0.9$ if 0.1 was sampled in the forward step or it sets $\epsilon = 0.1$ if 0.9 was sampled in the forward step.

Let me know if that works!

File: src/plugins/backward_sampling_schemes.jl [hunk 8]
```diff
@@ -10,7 +10,7 @@ Backward sampler that returns all noises of the corresponding node.
...
 struct CompleteSampler <: AbstractBackwardSamplingScheme end
 
-sample_backward_noise_terms(::CompleteSampler, node) = node.noise_terms
+sample_backward_noise_terms(::CompleteSampler, node::Node) = node.noise_terms
```
Comment (odow) [inline src/plugins/backward_sampling_schemes.jl:13]:
```suggestion
sample_backward_noise_terms(::CompleteSampler, node) = node.noise_terms
```

File: src/plugins/headers.jl [hunk 9]
```diff
@@ -115,10 +115,28 @@ abstract type AbstractBackwardSamplingScheme end
...
+        state::Dict{Symbol,Float64},
+    )::Vector{Noise}
+
+Returns a `Vector{Noise}` of noises sampled conditionally on the `state` using
+`sampler`.
+"""
+function sample_backward_noise_terms_with_state(
+    sampler::AbstractBackwardSamplingScheme,
+    node::Node,
+    state::Dict{Symbol,Float64},
```
Comment (odow) [inline src/plugins/headers.jl:135]:
```suggestion
   ::Dict{Symbol,Float64},
```

File: src/plugins/headers.jl [hunk 10]
```diff
@@ -115,10 +115,28 @@ abstract type AbstractBackwardSamplingScheme end
...
+        state::Dict{Symbol,Float64},
+    )::Vector{Noise}
+
+Returns a `Vector{Noise}` of noises sampled conditionally on the `state` using
+`sampler`.
+"""
+function sample_backward_noise_terms_with_state(
+    sampler::AbstractBackwardSamplingScheme,
+    node::Node,
+   ::Dict{Symbol,Float64},
```
Comment (odow) [inline src/plugins/headers.jl:135]:
```suggestion
    ::Dict{Symbol,Float64},
```

**Pull request merged**